### ISSUE TITLE
1.8.2 compiler crashes matching against map value with string interpolation

### ISSUE THREAD

**Issue by garthk** [elixir-lang/elixir#9257]
### Precheck

- [x] not a feature
- [x] searched for previous reports of the bug
- [x] not a security vulnerability
- [x] in a friendly mood

### Environment

```
$ uname
Darwin

$ elixir --version
Erlang/OTP 21 [erts-10.3.5.4] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe]

Elixir 1.8.2 (compiled with Erlang/OTP 21)
```

### Current behavior

The compiler crashes given my awful code trying to match from within string interpolation:

```elixir
defmodule M do
  def r do
    q = 1
    fn env -> %{"..." => "#{^q}"} = env end
  end
end
```

There's no need to save it to a file for a demo; just paste it into the `iex` prompt. 

If you're running `asdf`, a `.tool-versions` with `elixir 1.8.2-otp-21` does the trick. If you've Docker handy, paste it into `docker run --rm -ti elixir:1.8.2`. 

The error is:

```
** (FunctionClauseError) no function clause matching in :v3_core.pattern/2    
```

The only mention I found of that is:

> `** (FunctionClauseError) no function clause matching in :v3_core.pattern/2` is a compiler bug. Please open up a bug report with a minimal case, including the Elixir and Erlang/OTP versions! Thanks
>  – [José Valim](https://stackoverflow.com/q/56687751/5700#comment99963286_56687751)

### Expected behavior

An error message pointing at my awful code.

**Comment by garthk** [comment]
1.9 is much more helpful:
```
$ docker run --rm -ti elixir:1.9
Erlang/OTP 22 [erts-10.4.4] [source] [64-bit] [smp:2:2] [ds:2:2:10] [async-threads:1] [hipe]

Interactive Elixir (1.9.1) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> defmodule M do
...(1)>   def r do
...(1)>     q = 1
...(1)>     fn env -> %{"..." => "#{^q}"} = env end
...(1)>   end
...(1)> end
** (CompileError) iex:4: cannot invoke remote function String.Chars.to_string/1 inside a match
    (elixir) src/elixir_bitstring.erl:150: :elixir_bitstring.do_expand_expr/4
    (elixir) src/elixir_bitstring.erl:27: :elixir_bitstring.expand/8
    (elixir) src/elixir_bitstring.erl:10: :elixir_bitstring.expand/4
    (stdlib) lists.erl:1354: :lists.mapfoldl/3
    (stdlib) lists.erl:1355: :lists.mapfoldl/3
```
I'll leave this issue open in case you back-port this kind of bug, expecting you'll close it if you don't.

**Comment by garthk** [comment]
Excuse the noise; just noticed the `^` isn't necessary.

**Comment by josevalim** [comment]
We won’t backport burg fixes to v1.8 once v1.9 is out (except for security ones). Thanks!
